### PR TITLE
[tracer:context] introduce context object

### PR DESCRIPTION
**Pull request by ufoot**
Do the same we do in Python, use a context to propagate parent, current trace, and have this context overridable with an instrumentation-dependant provider.

Note: PR still needs polishing (tests & comments) but the basics are here.

### CODE REVIEW

File: test/tracer_test.rb [hunk 1]
```diff
@@ -142,4 +142,135 @@ def test_default_service
...
+
+    refute_equal(a.trace_id, b.trace_id, 'a and b do not belong to the same trace')
+    refute_equal(b.trace_id, c.trace_id, 'b and c do not belong to the same trace')
+    assert_equal(a.trace_id, c.trace_id, 'a and c belong to the same trace')
+    assert_equal(a.span_id, c.parent_id, 'a is the parent of c')
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def test_start_span_child_of_context
+    # skipping this for Ruby 1.9, as it has some threading issues (segfaults on msgpack)
```
Comment (ufoot) [inline test/tracer_test.rb]: OK, this sounds scary but it's really really a corner case I hit when using Ruby 1.9. Most of the code running runs fine, as long as you don't play with the fire and mix threads. What's happening here is that I'm submitting data in the main thread, that was created in another thread, and this causes a segfault with msgpack. And, as a side note, a Ruby version for which support has been dropped in years. So forwarding the problem upstream won't do any good. One alternative would be to have another context provider, that does not uses threads but another carrier, and do this test with it. I would typically rewrite this test when such a context provider is implemented for real.

File: lib/ddtrace/utils.rb [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
 module Datadog
-  # TODO[manu]: write docs
+  # Utils contains low-level utilities, typically to provide pseudo-random trace IDs.
   module Utils
+    # We use a custom random number generator because we want no interference
+    # with the default one. Using the default prng, we could break code that
+    # would rely on srand/rand sequences.
+    @rnd = Random.new
+    # We want to garantee all calls are sequential else, in some edge cases,
+    # the same value could be return twice, ending in trace collision.
```
Comment (ufoot) [inline lib/ddtrace/utils.rb]: I'm not 100.00% sure of this, could be worth a check, but it's hard to prove that you have a collision. I wrote that initially because I had some collisions, but the collisions were coming from elsewhere. I still do think it's relevant, but at that point I could also remove it. Taking any suggestion.

File: lib/ddtrace/utils.rb [hunk 3]
```diff
@@ -1,9 +1,21 @@
+require 'thread'
+
 module Datadog
-  # TODO[manu]: write docs
+  # Utils contains low-level utilities, typically to provide pseudo-random trace IDs.
   module Utils
+    # We use a custom random number generator because we want no interference
+    # with the default one. Using the default prng, we could break code that
+    # would rely on srand/rand sequences.
+    @rnd = Random.new
```
Comment (ufoot) [inline lib/ddtrace/utils.rb]: We had some issues with Go, here we do not have the same problem, but still, as stated in comment, we could break user code.

File: lib/ddtrace/span.rb [hunk 4]
```diff
@@ -39,15 +41,17 @@ def initialize(tracer, name, options = {})
...
       @meta = {}
       @metrics = {}
       @status = 0
 
       @parent = nil
       @sampled = true
 
-      @start_time = Time.now.utc
-      @end_time = nil
+      @start_time = nil # set by Tracer.start_span
```
Comment (ufoot) [inline lib/ddtrace/span.rb]: BREAKING CHANGE! Well, actually no, would break code relying on undocumented features and doing strange things, but this is still significant. I strongly believe is better to leave this unset, as this is a `start` time and not an `init` time. If it's not started, `start_time` is `nil`, period.

File: lib/ddtrace/span.rb [hunk 5]
```diff
@@ -91,19 +95,44 @@ def set_error(e)
...
     end
 
-    # Proxy function that flag a span as finished with the given
-    # timestamp. This function is used for retro-compatibility.
-    # DEPRECATED: remove this function in the next release
-    def finish_at(finish_time)
-      finish(finish_time)
-    end
+    # BREAKING CHANGE: finish_at has removed.
+    #
```
Comment (ufoot) [inline lib/ddtrace/span.rb]: Next step would be: remove this comment. But I would still leave it for this release, so that if someone was relying on it, it's just so obvious it disappeared (without browsing Github history).

Comment (ufoot) [inline lib/ddtrace/span.rb]: Oh, BTW, typo above... "`has been removed`" ...

File: lib/ddtrace/tracer.rb [hunk 6]
```diff
@@ -160,22 +222,9 @@ def set_tags(tags)
...
-
-      @tags.each { |k, v| span.set_tag(k, v) } unless @tags.empty?
-
-      # sampling
-      if parent.nil?
-        @sampler.sample(span)
-      else
-        span.sampled = span.parent.sampled
-      end
+      span = start_span(name, options)
```
Comment (ufoot) [inline lib/ddtrace/tracer.rb]: At some point `trace` method has almost no logic left in, compared to the Python impl I moved https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/tracer.py#L220 `get_call_context` to `start_span`. I think it's better this way but I'm worried I missed something.

File: lib/ddtrace/context.rb [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    # ThreadLocalContext can be used as a tracer global reference to create
+    # a different \Context for each thread. In synchronous tracer, this
+    # is required to prevent multiple threads sharing the same \Context
+    # in different executions.
+    def initialize
+      self.local = Datadog::Context.new
+    end
+
+    # Override the thread-local context with a new context.
+    def local=(ctx)
```
Comment (ufoot) [inline lib/ddtrace/context.rb]: In Python this is a couple of `get/set` functions, I think in Ruby overloading `=()` is better for that setter/getter pattern. Additionally, while writing the code, I was confused by getting/setting too many things. At least this is different from `Context.get` (the difference between `get` on `ThreadLocalContext and` `get` on `Context` which returns a tuple `traces, sampled` was making things a little unclear).

File: test/helper.rb [hunk 8]
```diff
@@ -63,36 +70,61 @@ def get_adapter_name
...
   def spans
-    spans = @spans
-    @spans = []
-    spans.flatten
+    @mutex.synchronize do
+      spans = @spans
+      @spans = []
+      spans.flatten!
+      # sort the spans to avoid test flakiness
+      spans.sort! do |a, b|
```
Comment (ufoot) [inline test/helper.rb]: I hesitated a lot before introducing this. Because in many cases it puts your spans upside down. I considered sorting them based on `start_time` first. But then in multtithreaded/concurrent context this is not garanteed. And these are the complex cases we want to sort out. So I ended up with that `ORDER BY name, resource, start_time, end_time` plan.

Comment (palazzem) [inline test/helper.rb]: Oh ok, removed previous comments because I wasn't aware of that. This is a test so everything that is fine to make testing easier and reliable 👍

File: test/tracecountheader_test.rb [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
 
   def setup
-    @server = WEBrick::HTTPServer.new Port: TEST_PORT
+    @log_buf = StringIO.new
+    log = WEBrick::Log.new @log_buf
+    access_log = [
+      [@log_buf, WEBrick::AccessLog::COMBINED_LOG_FORMAT]
+    ]
+
+    @server = WEBrick::HTTPServer.new Port: TEST_PORT, Logger: log, AccessLog: access_log
```
Comment (palazzem) [inline test/tracecountheader_test.rb]: I assume this one removes all the output when launching tests right?

File: test/tracer_test.rb (hunk 1 above)
Comment (palazzem) [inline test/tracer_test.rb]: Mhhh that's scary. What is the real problem here? are we breaking some support with Ruby 1.9? Also why msgpack is involved? it happens the same if we use the legacy mode (JSON?). Just to know what is happening here. Also if there is nothing we can do, I think it's better to expand the comment.

Nitpick - I think you can use that one instead of `return`:
```ruby
skip "Not available in Ruby 1.9" if RUBY_VERSION < '2.0.0'
```
otherwise the test exits with a success.

File: lib/ddtrace/utils.rb (hunk 3 above)
Comment (ufoot) [inline lib/ddtrace/utils.rb]: Calling the generator does alter it because it pulls out a number from the sequence. Example below:

```
srand 1234
puts [ rand, rand ] 
# 0.1915194503788923
# 0.6221087710398319

srand 1234
puts [ rand, rand ]
# 0.1915194503788923
# 0.6221087710398319

srand 1234
# do some tracing without this patch
puts [ rand, rand ]
# unexpected behavior (code relying on srand being predictive is broken)
```

Comment (palazzem) [inline lib/ddtrace/utils.rb]: ok in that case we can keep a separated `Random` instance so:
* our gem doesn't interfere with the sequence of the main program
* our sequence of random numbers is reproducible (say per thread)

File: lib/ddtrace/utils.rb (hunk 2 above)
Comment (ufoot) [inline lib/ddtrace/utils.rb]: I should dig dipper but reading this https://github.com/ruby/ruby/blob/trunk/random.c#L183 I see no protection whatsoever from running this in parallel. You don't  necessarily need to lock to get a random number but if you want to re-use the same generator across different threads -> yes you need. And this is what you'd do by using the default `rand`. But (my bad) as this is a C impl, it should be protected by the GIL so the collision is truely impossible (unless the GIL is removed, but this is not going to happen soon, I hear it could be planned for Ruby 3, but no hurry). Will remove the lock & leave a comment.

File: lib/ddtrace/tracer.rb (hunk 6 above)
Comment (ufoot) [inline lib/ddtrace/tracer.rb]: OK, got it. Will write a test with the above and comment the implementation to make this obvious. I think I just did not get it in the first place (the fact that `trace()` and `start_span()` had this by design difference about acquiring the context or not).

File: lib/ddtrace/utils.rb (hunk 2 above)
Comment (palazzem) [inline lib/ddtrace/utils.rb]: in real multi-threading languages of course you need a lock for a shared memory access, but the GIL ensures no collisions. About removing the GIL in Ruby 3 we'll see, since it's definitely something really hard to remove and that breaks any Ruby application.

File: lib/ddtrace/tracer.rb (hunk 6 above)
Comment (palazzem) [inline lib/ddtrace/tracer.rb]: Great! To be complete, at some point we'll end with (nomenclature is different but just to make it readable without "context"):
in the end we should end having 3 different ways:
* `start_active_span()`: uses context propagation and sets the span as active
* `start_manual_span()`: uses context propagation but doesn't set the active span
* `start_manual_span(ignore_active_span: true)`: doesn't use the context propagation and doesn't set the active span

For now the safest way is to use `start_span()` as a manual way, so that when we introduce these three working modes, `start_span()` will only be an alias that we can deprecate.

File: docs/GettingStarted.md [hunk 10]
```diff
@@ -583,6 +612,34 @@ On the server:
...
+for a ``tracegen`` service:
+
+    require 'ddtrace'
+
+    loop do
+	    Datadog.tracer.trace('hello-world') do |span|
+		    span.service = 'tracegen'
+		    span.resource = 'ruby'
+	    end
+	    sleep 1
```
Comment (palazzem) [inline docs/GettingStarted.md]: ah ok if that was the goal; the point is that usually people creates a span using a `sleep 1` to see something in the UI, and having the sleep inside our instrumentation block is the hello world everyone is going to use.

File: lib/ddtrace/utils.rb (hunk 3 above)
Comment (ufoot) [inline lib/ddtrace/utils.rb]: @palazzem our sequence of random numbers won't be reproducible (even, per thread) as all threads will use the same generator. But we don't care, all we need is no collision, our code does not rely on peculiar sequences of IDs, the more "pseudo-random", the better. But we definitely need to *not* interfere with the sequence of the main program.

Comment (palazzem) [inline lib/ddtrace/utils.rb]: @ufoot yes my point is generic and related to a testing environment where the sequence can be reproducible

File: test/tracecountheader_test.rb (hunk 9 above)
Comment (ufoot) [inline test/tracecountheader_test.rb]: Exactly, else tests where polluted by random Webrick output, not really relevant. Can be useful when debugging the test, else, useless.

File: lib/ddtrace/tracer.rb [hunk 11]
```diff
@@ -132,6 +144,68 @@ def set_tags(tags)
       @tags.update(tags)
     end
 
+    # Guess context and parent from child_of entry.
+    def guess_context_and_parent(options = {})
```
Comment (palazzem) [inline lib/ddtrace/tracer.rb]: 👍  since it's a private method
